src/librustc_typeck/check/closure.rs

```diff
@@ -269,4 +273,408 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+///
+/// - First, each type R only contains **bound regions that
+///   originated from the corresponding expected type E**.
+///   - That is, if the user wrote `|x: &str, y|`, the resulting bound
+///     region name in `x` (`Anon(0)`) would not be combined with some
+///     bound region name that appears in the expected type for `y`. The bound region
+///     names from the expected/supplied types effectively inherit distinct
+///     namespaces, and should never be mixed.
+/// - Second, any bound region appears that appears in some expected type E,
+///   also appears in the corresponding spot in the resulting type R.
+///   - In particular, imagine that we have an expected signature
+///     `(&str) -> &str`, but the user wrote `|x: &'a str| ...`. In
+///     the supplied type, the region `'a` is free.
+///     - We could choose to use the final signature `(&'a str) ->
+///       &str` without violating the first invariant, since we did
+///       "intermix" any bound region names from the supplied types
+///       with those of the expected type.
+///     - However, if we did so, we have lost the constraint expressed
+///       in the expected signature, that the return type must relate
+///       to the input.  In fact, this resulting signature is
+///       inconsistent, since a bound region cannot only appear in the
+///       input (this was #38714).
+///
+/// To produce the type R, we relate the types recursively, which
+/// requires that the types generally match. If that is not true, an
+/// error results, which will result in the expected signature being
+/// ignored. The places where the types do not have to match, and the
+/// resulting type in those scenarios, are as follows:
+///
+/// - If the expected type E is an (unbound) inference variable,
+///   and S is not an inference variable, then the resulting type will be S,
+///   so long as S does not contain anonymous regions.
+/// - If the supplied type S is an unbound inference variable (i.e.,
+///   the user wrote `_`), we will use E.
+/// - When we encounter types with internal binders, e.g. `fn(&str)`
+///   or `dyn Fn(&str)`, then we want to take the supplied type
+///   (justification below). To preserve the invariants listed above,
+///   we will error if either of the following is true (but otherwise
+///   use the supplied type):
+///   - the supplied type includes any regions bound at the closure depth
+///     - in this case, using the supplied type would inject names from the supplied
+///       side into the closure binder
+///   - the expected type includes any regions bound at the closure depth
+///     - in this case, using the supplied type would **remove** names from the closure
+///       binder that the expected type included
+///
+/// Regions are allowed to differ as follows:
+///
+/// - If S supplies a late-bound region bound at the closure depth (1):
+///   - If that region is anonymous, we use the region from E (which may also be free).
```

Comment:
This only works because closures don't do elision e.g. `|x: &u32| -> &u32` is `for<'a> |x: &'a u32| -> &'_ u32` rather than `for<'a> |x: &'a u32| -> &'a u32`. Worth noting and testing.

Reply:
Woah. I .. didn't realize this was the case actually! Doesn't that seem like a bug? I at least expect `|x: &u32| -> &u32` in a closure signature to mean the same as `fn(&u32) -> &u32`.

(Also, `'_` in that position *does* do elision, as far as I know, but I suspect you mean "fresh region"? If so, hopefully a fresh *free* region, since a fresh *bound* region in that position would be broken.)